### PR TITLE
Refactor/user function handling modules +  Manager can run additional worker on thread

### PR DESCRIPTION
**Pull request by jlnav**
Addresses #1166 

# User - Manager runs local worker on thread - Gen Worker 0

The manager can start an additional worker _thread_ with the worker ID 0. 

This worker is, via `W["gen_worker"] = True`, and filtering options, made to run gens only.

Tentative interface: `libE_specs["gen_on_manager"] = True`

# Internal - Refactor runners.py

Largely refactors utils/runners.py so additional "Runners" for user functions can be more easily developed/extended.

GlobusComputeRunner and ThreadRunner inherit from Runner.

Runner inspects the contents of the input sim/gen specs and instantiates the correct subclass on the left-hand-side:

```
>>> sim_specs = {... , "globus_compute_endpoint": "1234}
>>> runner = Runner(sim_specs)
>>> runner
<libensemble.utils.runners.GlobusComputeRunner object at 0x1490ec190>
>>> runner = Runner({"sim_f": print})
>>> runner
<libensemble.utils.runners.Runner object at 0x104e024d0>
```

# Internal -  _WorkerIndexer class for indexing into worker arrays

# Internal - Workers can be assigned gen_worker.

### CODE REVIEW

File: libensemble/manager.py [hunk 1]
```diff
@@ -154,6 +155,48 @@ def filter_nans(array: npt.NDArray) -> npt.NDArray:
 """
 
 
+class _Worker:
+    """Wrapper class for Worker array and worker comms"""
+
+    def __init__(self, W: npt.NDArray, wid: int, wcomms: list = []):
+        self.__dict__["_W"] = W
```
Comment (shuds13) [inline libensemble/manager.py:162]: How different from self._W = W

**Comment by shuds13** [comment]
From a quick look, I don't understand the number of calls to the _Worker constructor. Dont we just need to create one wrapper class? This could use the __getitem__ method to set indexing. e.g.

class _Worker:
    def __init__(self, W, manager_as_worker=False):
        self.data = data
        self.manager_as_worker = manager_as_worker

    def __getitem__(self, key):
        if self.manager_as_worker:
            return self.data[key]
        else:
            return self.data[key - 1]

**Comment by jlnav** [comment]
Yeah, I don't like the number of calls myself to the _Worker constructor. I was leaning towards something like in manager.run:

```python
for w in Work:
    if self._sim_max_given():
        break
  w = _Worker(
  self._check_work_order(Work[w], w)
  ...
```

But avoided it at the time since I didn't want to rewrite too much logic also dealing with passing worker IDs around e.g.:

```python
            if w in self.persis_pending:
                self.persis_pending.remove(w)
                worker.active = 0
            self._freeup_resources(w)
```

So you're imagining the interface being something like `self.worker[w].active = 0` then? Yeah, the dynamic indexing would still work and so would the attribute setting/getting, plus the same class wouldn't have to be made multiple times.

**Comment by shuds13** [comment]
You'll see in #1142 I made a change in worker_resources to allow assigning resources to the manger.

**Comment by jlnav** [comment]
TODO on additional branch:

- [ ] Ensure resources are correctly allocated, especially considering "resources" objects being identical for the manager/worker_0 as they're in the same interpreter
- [ ] Check dumped_every_k persis_info matches that maintained by worker 0 within a persistent gen

**Comment by shuds13** [comment]
Its debatable but I'm inclined to make use of extra worker transparent to user. Maybe "gen_on_thread" or something. Not "manager_runs_additional_worker".

File: libensemble/manager.py [hunk 2]
```diff
@@ -208,6 +230,34 @@ def __init__(
             (1, "stop_val", self.term_test_stop_val),
         ]
 
+        if self.libE_specs.get("manager_runs_additional_worker", False):
```
Comment (shuds13) [inline libensemble/manager.py:233]: I would think should be an "else" here.

Comment (jlnav) [reply libensemble/manager.py:233]: I'm not sure what would go in "else", though. We do extra setup for an additional worker, but there's nothing alternative to do otherwise.

Comment (shuds13) [reply libensemble/manager.py:256]: The lines that are always done above
        else:
            self.W = np.zeros(len(self.wcomms), dtype=Manager.worker_dtype)
            self.W["worker_id"] = np.arange(len(self.wcomms)) + 1

File: libensemble/comms/comms.py [hunk 3]
```diff
@@ -146,10 +146,11 @@ def mail_flag(self):
 
 
 class QCommLocal(Comm):
-    def __init__(self, main, nworkers, *args, **kwargs):
+    def __init__(self, main, *args, **kwargs):
         self._result = None
         self._exception = None
         self._done = False
+        self._ufunc = kwargs.get("ufunc", False)
```
Comment (shuds13) [inline libensemble/comms/comms.py:153]: self._ufunc appears not to be used currently

File: libensemble/comms/comms.py [hunk 4]
```diff
@@ -208,10 +209,13 @@ def result(self, timeout=None):
         return self._result
 
     @staticmethod
-    def _qcomm_main(comm, main, *args, **kwargs):
+    def _qcomm_main(comm, main, *fargs, **kwargs):
```
Comment (shuds13) [inline libensemble/comms/comms.py:212]: I would probably keep as *args by convention

File: libensemble/comms/comms.py [hunk 5]
```diff
@@ -208,10 +209,13 @@ def result(self, timeout=None):
...
         """Main routine -- handles return values and exceptions."""
         try:
-            _result = main(comm, *args, **kwargs)
+            if not kwargs.get("ufunc"):
```
Comment (shuds13) [inline libensemble/comms/comms.py:215]: Call something like user_func to not confuse with ufunc for universal functions.

File: libensemble/manager.py [hunk 6]
```diff
@@ -597,7 +652,7 @@ def _alloc_work(self, H: npt.NDArray, persis_info: dict) -> dict:
 
         alloc_f = self.alloc_specs["alloc_f"]
         output = alloc_f(
-            self.W,
+            self.W.iterable,
```
Comment (shuds13) [inline libensemble/manager.py:658]: I'm uneasy about this. Should we not use self.W in the same way in alloc funcs (inc. in alloc support). Esp. as there is potentially invisible errors if you end up with W[-1]

Comment (jlnav) [reply libensemble/manager.py:658]: Good point. I'll refactor more of alloc_support

File: libensemble/tools/alloc_support.py [hunk 7]
```diff
@@ -87,29 +87,22 @@ def assign_resources(self, rsets_req, use_gpus=None, user_params=[]):
             rset_team = self.sched.assign_resources(rsets_req, use_gpus, user_params)
         return rset_team
 
-    def avail_worker_ids(self, persistent=None, active_recv=False, zero_resource_workers=None):
+    def avail_worker_ids(self, persistent=False, active_recv=False, zero_resource_workers=None, worker_type=None):
```
[the remark this answers is not in the document]
Comment (jlnav) [reply libensemble/tools/alloc_support.py:90]: When we can filter for specific types of workers, gen workers or sim workers, then their persistent state doesn't also need to specify persistent_sim or persistent_gen. There's no situation where a gen worker should have a persistent_sim state, so its best that "persistent" describes the worker as a bool

File: libensemble/manager.py (hunk 6 above)
Comment (jlnav) [reply libensemble/manager.py:658]: One complication I just realized is that that'll expose the _WorkerIndexer wrapper to the user. `W` as traditionally within the alloc_f would no longer be be a NumPy array unless we pass in the iterable as we do here.

Comment (jlnav) [reply libensemble/manager.py:658]: Ah, but if we don't explicitly use W outside of persistent_support functions, then we can wrap it with an indexer inside PersistentSupport

File: libensemble/tools/alloc_support.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (shuds13) [reply libensemble/tools/alloc_support.py:90]: None means give me a list of all workers, persistent or not.

[the remark this answers is not in the document]
Comment (shuds13) [reply libensemble/tools/alloc_support.py:90]: Perhaps we don't use it. But we should agree on it.

File: libensemble/manager.py (hunk 6 above)
Comment (shuds13) [reply libensemble/manager.py:639]: I'm leaning towards we pass self.W, even if its breaking. I will review the alloc funcs.

File: libensemble/manager.py [hunk 8]
```diff
@@ -158,13 +159,36 @@ class Manager:
 
     worker_dtype = [
         ("worker_id", int),
+        ("worker_type", int),
         ("active", int),
-        ("persis_state", int),
-        ("active_recv", int),
+        ("persistent", bool),
```
Comment (shuds13) [inline libensemble/manager.py:164]: Are these changes necessary?

**Review by shuds13 — CHANGES_REQUESTED** [review]
What I would do:
 - Make sure W changes are consistent (either 3 booleans + worker type or as it was)
 - persistent (default changed to False rather than None - how does this affect things.)
 - Make gen_worker field for W - mirrors zero_resource_worker field.
 - Pass self.W to alloc and modify as necessary (I think).
 - testing
   - test giving gen diff resources.

File: libensemble/tools/alloc_support.py [hunk 9]
```diff
@@ -87,24 +87,19 @@ def assign_resources(self, rsets_req, use_gpus=None, user_params=[]):
             rset_team = self.sched.assign_resources(rsets_req, use_gpus, user_params)
         return rset_team
 
-    def avail_worker_ids(self, persistent=None, active_recv=False, zero_resource_workers=None):
+    def avail_worker_ids(self, persistent=None, active_recv=False, zero_resource_workers=None, gen_workers=False):
```
Comment (shuds13) [inline libensemble/tools/alloc_support.py:90]: Shouldn't gen_workers default to None, like zero_resource_workers?

File: libensemble/tools/alloc_support.py [hunk 10]
```diff
@@ -87,24 +87,19 @@ def assign_resources(self, rsets_req, use_gpus=None, user_params=[]):
...
         """Returns available workers as a list of IDs, filtered by the given options.
 
         :param persistent: (Optional) Int. Only return workers with given ``persis_state`` (1=sim, 2=gen).
         :param active_recv: (Optional) Boolean. Only return workers with given active_recv state.
         :param zero_resource_workers: (Optional) Boolean. Only return workers that require no resources.
+        :param gen_workers: (Optional) Boolean. If True, return gen-only workers and manager's ID.
```
Comment (shuds13) [inline libensemble/tools/alloc_support.py:96]: why return managers ID separatly is that necessary

**Comment by jlnav** [comment]
> I think the only issue now, is thread-lock not getting cleaned up between libE() calls.

This is fixed. I'll merge when approved.

File: docs/data_structures/libE_specs.rst [hunk 11]
```diff
@@ -28,7 +28,11 @@ libEnsemble is primarily customized by setting options within a ``LibeSpecs`` cl
                     Manager/Worker communications mode: ``'mpi'``, ``'local'``, or ``'tcp'``.
 
                 **nworkers** [int]:
-                    Number of worker processes in ``"local"`` or ``"tcp"``.
+                    Number of worker processes in ``"local"``, ``"threads"``, or ``"tcp"``.
```
Comment (jmlarson1) [inline docs/data_structures/libE_specs.rst:31]: but not `mpi`?

File: libensemble/alloc_funcs/only_one_gen_alloc.py [hunk 12]
```diff
@@ -21,22 +21,25 @@ def ensure_one_active_gen(W, H, sim_specs, gen_specs, alloc_specs, persis_info,
     gen_flag = True
     gen_in = gen_specs.get("in", [])
 
-    for wid in support.avail_worker_ids():
-        persis_info = support.skip_canceled_points(H, persis_info)
-
-        if persis_info["next_to_give"] < len(H):
+    if persis_info["next_to_give"] < len(H):
+        for wid in support.avail_worker_ids(gen_workers=False):
+            persis_info = support.skip_canceled_points(H, persis_info)
```
Comment (jmlarson1) [inline libensemble/alloc_funcs/only_one_gen_alloc.py:26]: Do we only want to skip canceled points if the above two conditions are satisfied?

File: docs/data_structures/libE_specs.rst (hunk 11 above)
Comment (jlnav) [reply docs/data_structures/libE_specs.rst:31]: I don't believe the nworkers option has ever influenced mpi comms, only `mpirun -n N` makes a difference

File: libensemble/alloc_funcs/only_one_gen_alloc.py (hunk 12 above)
Comment (jlnav) [reply libensemble/alloc_funcs/only_one_gen_alloc.py:26]: The one condition?

**Comment by shuds13** [comment]
@jlnav 

- In start_only_persistent why not gen_worker=False on the sim
- In inverse bayes - should H["gen_worker"] by gen_workers? (if so we need a test).
- What about fast_alloc_and_pausing and give_pregenerated_work?

**Review by shuds13 — APPROVED** [review]
